### PR TITLE
UCT/MEM/V2: updated docs

### PR DESCRIPTION
**Pull request by hoopoepg**
- updated comments on functions descriptions

### CODE REVIEW

File: src/uct/api/v2/uct_v2.h [hunk 1]
```diff
@@ -135,10 +135,10 @@ typedef enum {
 
 typedef enum {
     /**
-     * Invalidate memory region. If this flag is set then memory region is
-     * invalidated after de-registration and callback (see @ref
+     * Invalidate the memory region. If this flag is set then memory region is
+     * invalidated after de-registration and the callback (see @ref
      * uct_md_mem_dereg_params_t) is is called when the memory is fully
-     * invalidated and would not be accessed anymore by zero-copy or remote
+     * invalidated and will not be accessed any more by zero-copy or remote
```
Comment (brminich) [inline src/uct/api/v2/uct_v2.h:141]: please remove double is on line 140

Comment (hoopoepg) [reply src/uct/api/v2/uct_v2.h]: yep, fixed

File: src/uct/api/v2/uct_v2.h [hunk 2]
```diff
@@ -135,10 +135,10 @@ typedef enum {
...
 typedef enum {
     /**
-     * Invalidate memory region. If this flag is set then memory region is
-     * invalidated after de-registration and callback (see @ref
-     * uct_md_mem_dereg_params_t) is is called when the memory is fully
-     * invalidated and would not be accessed anymore by zero-copy or remote
+     * Invalidate the memory region. If this flag is set then memory region is
+     * invalidated after de-registration and the callback (see @ref
+     * uct_md_mem_dereg_params_t) is called when the memory is fully
+     * invalidated and will not be accessed any more by zero-copy or remote
```
Comment (brminich) [inline src/uct/api/v2/uct_v2.h:141]: `anymore` was correct, afaik `any more` has different meaning than what was meant here

**Review by brminich — APPROVED** [review]

**Comment by shamisp** [comment]
Is this follow up on https://github.com/openucx/ucx/pull/6819#pullrequestreview-668413762 ? 
Can you please push these changes to the original branch so Tony and I can review it ? Once we done over there it makes sense to open a PR.

**Comment by hoopoepg** [comment]
not sure are you what about, but here used same git branch

**Review by yosefe — APPROVED** [review]

**Comment by shamisp** [comment]
@hoopoepg  branch - odd. 

Anyways - there are still plenty of comments to address from https://github.com/openucx/ucx/pull/6819. 
The documentation of semantics of the callback invocation is unclear - what is fully invalidated ? 
Function documentation is missing as well. References to an actual implementation will help. Do you plan to PR implementation before 1.11 branch ?

**Review by shamisp — CHANGES_REQUESTED** [review]
Please address all comments from https://github.com/openucx/ucx/pull/6819

**Comment by hoopoepg** [comment]
> Please address all comments from #6819

this PR addresses comments from #6819

**Comment by shamisp** [comment]
> > Please address all comments from #6819
> 
> this PR addresses comments from #6819

it is incomplete. It does not answer question when the call back invoked. What is the invalidation semantics and when it happens ? Function documentation is incomplete (`@brief` is not enough), etc.  Please take a look at review comments.

**Comment by shamisp** [comment]
Based on the code it seams that invalidation is invoked when memory region is de-registered or evicted from rcache ?

**Comment by hoopoepg** [comment]
> evicted from rcache

this one is correct. rcache just notify backend about memory region is released in rcache and may be destroyed/deregistered

**Comment by shamisp** [comment]
> > evicted from rcache
> 
> this one is correct. rcache just notify backend about memory region is released in rcache and may be destroyed/deregistered

Well, if rcache is not used , it supposed to be invoked on reregistration >

**Comment by hoopoepg** [comment]
right, invoked after de-registration

**Comment by shamisp** [comment]
@hoopoepg Each one of the cases has to be explicitly documented

**Comment by hoopoepg** [comment]
@shamisp callback is called when memory region is released by rcache or directly from dereg function, no matter how is it implemented behind of this API, this is reflected in description of UCT_MD_MEM_DEREG_FLAG_INVALIDATE flag. for now rcache backend API is called as reg/dereg/dump, but this is just historical naming: it could be m[un]map, posix/sysv release, etc, but in any case callback is called after such API is called and memory domain API doesn't care about this. we can describe all possible cases how/when callback may be called, but all of them may be described as "it will be called when memory region is **really** released/deregistered".

**Comment by hoopoepg** [comment]
@shamisp I added notes for UCT_MD_MEM_DEREG_FLAG_INVALIDATE flag - described callback call ways

File: src/uct/api/v2/uct_v2.h [hunk 3]
```diff
@@ -212,7 +217,7 @@ uct_iface_estimate_perf(uct_iface_h tl_iface, uct_perf_attr_t *perf_attr);
  * @brief Undo the operation of @ref uct_md_mem_reg() and invalidate memory
  *        region.
  *
- * @param [in]  md          Memory domain which was used to register the memory.
```
Comment (shamisp) [inline src/uct/api/v2/uct_v2.h:215]: Please add full description in addition to brief description.

Comment (hoopoepg) [reply src/uct/api/v2/uct_v2.h:215]: added

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Comment by yosefe** [comment]
> @yosefe ok to merge?

@shamisp should approve

**Review by shamisp — APPROVED** [review]

**Comment by shamisp** [comment]
@hoopoepg please squash and we are good to go.

**Comment by brminich** [comment]
@tonycurtis, do you have additional comments?

File: src/uct/api/v2/uct_v2.h [hunk 4]
```diff
@@ -135,10 +135,10 @@ typedef enum {
 
 typedef enum {
     /**
-     * Invalidate memory region. If this flag is set then memory region is
-     * invalidated after de-registration and callback (see @ref
-     * uct_md_mem_dereg_params_t) is is called when the memory is fully
-     * invalidated and would not be accessed anymore by zero-copy or remote
+     * Invalidate the memory region. If this flag is set then memory region is
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:138]: + the memory region

File: src/uct/api/v2/uct_v2.h [hunk 5]
```diff
@@ -149,8 +149,13 @@ typedef enum {
  * @ingroup UCT_MD
  * @brief Completion callback for memory region invalidation.
  *
- * This callback routine is invoked when when it is no longer accessible by
+ * This callback routine is invoked when it is no longer accessible by
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:152]: ambiguous use of "it"

File: src/uct/api/v2/uct_v2.h [hunk 6]
```diff
@@ -149,8 +149,13 @@ typedef enum {
...
  * remote peer.
+ * 
+ * $note: in some implementations this callback may be called immediately after
+ *        @ref uct_md_mem_dereg_v2 is called, but it is possible that callback
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:156]: that the callback call will be delayed until all references to the memory region, for example, when ... in the implementation of the memory domains.

File: src/uct/api/v2/uct_v2.h [hunk 7]
```diff
@@ -212,7 +217,11 @@ uct_iface_estimate_perf(uct_iface_h tl_iface, uct_perf_attr_t *perf_attr);
  * @brief Undo the operation of @ref uct_md_mem_reg() and invalidate memory
  *        region.
  *
- * @param [in]  md          Memory domain which was used to register the memory.
+ * This routine deregisters memory region registered by @ref uct_md_mem_reg and
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:220]: the memory region

File: src/uct/api/v2/uct_v2.h [hunk 8]
```diff
@@ -212,7 +217,11 @@ uct_iface_estimate_perf(uct_iface_h tl_iface, uct_perf_attr_t *perf_attr);
...
+ * allow to invalidate memory region with callback called when memory region is
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:221]: the memory region again.

"allow the memory region to be invalidated..." ?

**Review by tonycurtis — CHANGES_REQUESTED** [review]

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
@tonycurtis ok to merge?

**Review by tonycurtis — APPROVED** [review]

**Pull request merged**